### ISSUE TITLE
Error running any xonsh script

### ISSUE THREAD

**Issue by kalekundert** [xonsh/xonsh#3055]
I'm getting a stack trace whenever I try to run any file as a xonsh script.  I have a feeling I'm doing something wrong, but I can't figure out what it is, so hopefully someone can set me on the right track:

```sh
$ touch foo.xsh
$ xonsh --no-rc foo.xsh
Traceback (most recent call last):
  File "/home/kale/.virtualenvs/python37/lib/python3.7/site-packages/xonsh/__amalgam__.py", line 24001, in main
    args = premain(argv)
  File "/home/kale/.virtualenvs/python37/lib/python3.7/site-packages/xonsh/__amalgam__.py", line 23952, in premain
    env = start_services(shell_kwargs, args)
  File "/home/kale/.virtualenvs/python37/lib/python3.7/site-packages/xonsh/__amalgam__.py", line 23887, in start_services
    cacheall=shell_kwargs.get("cacheall", False),
  File "/home/kale/.virtualenvs/python37/lib/python3.7/site-packages/xonsh/__amalgam__.py", line 23018, in __init__
    load_builtins(execer=self, ctx=xonsh_ctx)
  File "/home/kale/.virtualenvs/python37/lib/python3.7/site-packages/xonsh/__amalgam__.py", line 22589, in load_builtins
    builtins.__xonsh__.load(execer=execer, ctx=ctx)
  File "/home/kale/.virtualenvs/python37/lib/python3.7/site-packages/xonsh/__amalgam__.py", line 22665, in load
    self.env = Env(default_env())
  File "/home/kale/.virtualenvs/python37/lib/python3.7/site-packages/xonsh/__amalgam__.py", line 19107, in __init__
    self[key] = val
  File "/home/kale/.virtualenvs/python37/lib/python3.7/site-packages/xonsh/__amalgam__.py", line 19280, in __setitem__
    deval = ensurer.detype(val)
  File "/home/kale/.virtualenvs/python37/lib/python3.7/site-packages/xonsh/__amalgam__.py", line 5456, in detype
    return x.detype()
  File "/home/kale/.virtualenvs/python37/lib/python3.7/site-packages/xonsh/__amalgam__.py", line 18129, in detype
    style = self.style
  File "/home/kale/.virtualenvs/python37/lib/python3.7/site-packages/xonsh/__amalgam__.py", line 18156, in style
    style_name = self.style_name
  File "/home/kale/.virtualenvs/python37/lib/python3.7/site-packages/xonsh/__amalgam__.py", line 18146, in style_name
    env = builtins.__xonsh__.env
AttributeError: 'XonshSession' object has no attribute 'env'
Xonsh encountered an issue during launch
Failback to /bin/sh
sh: cannot set terminal process group (-1): Inappropriate ioctl for device
sh: no job control in this shell
sh-4.4$ 
```
This example is just executing an empty file, but I get the same stack trace with files containing commands that run fine in interactive mode, and also with files that don't exist.  I also get the same error with or without the `--no-rc` option, but I included it here in the hopes of limiting the scope of the problem.

## xonfig

```
$ xonfig
+------------------+----------------------+
| xonsh            | 0.8.12               |
| Git SHA          | 986bf28d             |
| Commit Date      | Mar 11 14:37:52 2019 |
| Python           | 3.7.0                |
| PLY              | 3.11                 |
| have readline    | True                 |
| prompt toolkit   | 1.0.15               |
| shell type       | readline             |
| pygments         | 2.2.0                |
| on posix         | True                 |
| on linux         | True                 |
| distro           | unknown              |
| on darwin        | False                |
| on windows       | False                |
| on cygwin        | False                |
| on msys2         | False                |
| is superuser     | False                |
| default encoding | utf-8                |
| xonsh encoding   | utf-8                |
| encoding errors  | surrogateescape      |
+------------------+----------------------+
```

**Comment by kalekundert** [comment]
Update: I was able to poke around at this a little more and understand the problem a little better.  First, I actually get the above stack trace any time I run `xonsh` at all, script or no, except when launching terminals via my window manager.  This exception made me think there was something wrong with my environment.

So I systematically eliminated variables from the environment, and found that I could run xonsh with the following command:
```
$ env -u LS_COLORS xonsh --no-rc

                      Welcome to the xonsh shell (0.8.12)                       

                              ~ None shell pass ~                               

--------------------------------------------------------------------------------
xonfig tutorial    ->    Launch the tutorial in the browser
xonfig wizard      ->    Run the configuration wizard and claim your shell 
(Note: Run the Wizard or create a ~/.xonshrc file to suppress the welcome screen)

kale@thinkpad ~ $ 
```

Furthermore, simply deleting `$LS_COLORS` also restores function:
```
$ del $LS_COLORS
$ xonsh
...
```

My `$LS_COLORS` variable is as follows:
```
$ echo $LS_COLORS
*.7z=1;0;31:*.Z=1;0;31:*.aac=0;36:*.ace=1;0;31:*.alz=1;0;31:*.arc=1;0;31:*.arj=1;0;31:*.asf=1;0;35:*.au=0;36:*.avi=1;0;35:*.bmp=1;0;35:*.bz=1;0;31:*.bz2=1;0;31:*.cab=1;0;31:*.cgm=1;0;35:*.cpio=1;0;31:*.deb=1;0;31:*.dl=1;0;35:*.dwm=1;0;31:*.dz=1;0;31:*.ear=1;0;31:*.emf=1;0;35:*.esd=1;0;31:*.flac=0;36:*.flc=1;0;35:*.fli=1;0;35:*.flv=1;0;35:*.gif=1;0;35:*.gl=1;0;35:*.gz=1;0;31:*.jar=1;0;31:*.jpeg=1;0;35:*.jpg=1;0;35:*.lha=1;0;31:*.lrz=1;0;31:*.lz=1;0;31:*.lz4=1;0;31:*.lzh=1;0;31:*.lzma=1;0;31:*.lzo=1;0;31:*.m2v=1;0;35:*.m4a=0;36:*.m4v=1;0;35:*.mid=0;36:*.midi=0;36:*.mjpeg=1;0;35:*.mjpg=1;0;35:*.mka=0;36:*.mkv=1;0;35:*.mng=1;0;35:*.mov=1;0;35:*.mp3=0;36:*.mp4=1;0;35:*.mp4v=1;0;35:*.mpc=0;36:*.mpeg=1;0;35:*.mpg=1;0;35:*.nuv=1;0;35:*.oga=0;36:*.ogg=0;36:*.ogm=1;0;35:*.ogv=1;0;35:*.ogx=1;0;35:*.opus=0;36:*.pbm=1;0;35:*.pcx=1;0;35:*.pgm=1;0;35:*.png=1;0;35:*.ppm=1;0;35:*.qt=1;0;35:*.ra=0;36:*.rar=1;0;31:*.rm=1;0;35:*.rmvb=1;0;35:*.rpm=1;0;31:*.rz=1;0;31:*.sar=1;0;31:*.spx=0;36:*.svg=1;0;35:*.svgz=1;0;35:*.swm=1;0;31:*.t7z=1;0;31:*.tar=1;0;31:*.taz=1;0;31:*.tbz=1;0;31:*.tbz2=1;0;31:*.tga=1;0;35:*.tgz=1;0;31:*.tif=1;0;35:*.tiff=1;0;35:*.tlz=1;0;31:*.txz=1;0;31:*.tz=1;0;31:*.tzo=1;0;31:*.tzst=1;0;31:*.vob=1;0;35:*.war=1;0;31:*.wav=0;36:*.webm=1;0;35:*.wim=1;0;31:*.wmv=1;0;35:*.xbm=1;0;35:*.xcf=1;0;35:*.xpm=1;0;35:*.xspf=0;36:*.xwd=1;0;35:*.xz=1;0;31:*.yuv=1;0;35:*.z=1;0;31:*.zip=1;0;31:*.zoo=1;0;31:*.zst=1;0;31:bd=40;1;0;33:ca=0;30;41:cd=40;1;0;33:di=1;0;34:do=1;0;35:ex=1;0;32:ln=1;0;36:mh=0:mi=0:or=40;1;0;31:ow=0;34;42:pi=40;0;33:rs=0:sg=0;30;43:so=1;0;35:st=0;37;44:su=0;37;41:tw=0;30;42
```

This value is set in my `~/.xonshrc`, and seems to works fine as long as it's not defined when xonsh is starting up.  It's still not clear to me what about this variable causes `xonsh` to crash on startup, but I suspect it has to do with the `LsColors` class in `environ.py`.

**Comment by scopatz** [comment]
Hi @kalekundert! Thanks for reporting, I'll try to take a look soon!

**Comment by scopatz** [comment]
Sorry for the long delay in getting back to this.  I cannot reproduce this issue.  Is it still a problem?

**Comment by Drummersbrother** [comment]
Hi! This is quite an old issue, but this problem has appeared during the last few months for me so I want to resurrect it.
My configuration is having the same issue, but deleting $LS_COLORS seems to fix it.
The traceback I got and my xonfig are below: 
```sh
+------------------+----------------------+
| xonsh            | 0.9.13.dev1          |
| Git SHA          | 9f7ccc65             |
| Commit Date      | Oct 15 17:14:50 2019 |
| Python           | 3.7.4                |
| PLY              | 3.11                 |
| have readline    | True                 |
| prompt toolkit   | 2.0.9                |
| shell type       | prompt_toolkit2      |
| pygments         | 2.4.2                |
| on posix         | True                 |
| on linux         | True                 |
| distro           | arch                 |
| on darwin        | False                |
| on windows       | False                |
| on cygwin        | False                |
| on msys2         | False                |
| is superuser     | False                |
| default encoding | utf-8                |
| xonsh encoding   | utf-8                |
| encoding errors  | surrogateescape      |
+------------------+----------------------+
```
 
```py
$ # Trying to run my script
$ xonsh setup_sql_proxy.xsh                                                                                           
Traceback (most recent call last):
  File "/usr/lib/python3.7/site-packages/xonsh/__amalgam__.py", line 24015, in main
    args = premain(argv)
  File "/usr/lib/python3.7/site-packages/xonsh/__amalgam__.py", line 23968, in premain
    env = start_services(shell_kwargs, args)
  File "/usr/lib/python3.7/site-packages/xonsh/__amalgam__.py", line 23903, in start_services
    cacheall=shell_kwargs.get("cacheall", False),
  File "/usr/lib/python3.7/site-packages/xonsh/__amalgam__.py", line 23029, in __init__
    load_builtins(execer=self, ctx=xonsh_ctx)
  File "/usr/lib/python3.7/site-packages/xonsh/__amalgam__.py", line 22760, in load_builtins
    builtins.__xonsh__.load(execer=execer, ctx=ctx)
  File "/usr/lib/python3.7/site-packages/xonsh/__amalgam__.py", line 22832, in load
    self.env = Env(default_env())
  File "/usr/lib/python3.7/site-packages/xonsh/__amalgam__.py", line 19255, in __init__
    self[key] = val
  File "/usr/lib/python3.7/site-packages/xonsh/__amalgam__.py", line 19428, in __setitem__
    deval = ensurer.detype(val)
  File "/usr/lib/python3.7/site-packages/xonsh/__amalgam__.py", line 5498, in detype
    return x.detype()
  File "/usr/lib/python3.7/site-packages/xonsh/__amalgam__.py", line 18217, in detype
    style = self.style
  File "/usr/lib/python3.7/site-packages/xonsh/__amalgam__.py", line 18249, in style
    style_name = self.style_name
  File "/usr/lib/python3.7/site-packages/xonsh/__amalgam__.py", line 18239, in style_name
    env = builtins.__xonsh__.env
AttributeError: 'XonshSession' object has no attribute 'env'
Xonsh encountered an issue during launch
Failback to /bin/sh
sh: cannot set terminal process group (-1): Inappropriate ioctl for device
sh: no job control in this shell
sh-5.0$
```

**Comment by worldmind** [comment]
I have not the same, but something like when I try to run py.test:
```
py.test -v additions/scripts/test_book.py                                                                                     
INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>   File "/home/worldmind/.local/lib/python3.7/site-packages/_pytest/main.py", line 189, in wrap_session
INTERNALERROR>     config._do_configure()
INTERNALERROR>   File "/home/worldmind/.local/lib/python3.7/site-packages/_pytest/config/__init__.py", line 717, in _do_configure
INTERNALERROR>     self.hook.pytest_configure.call_historic(kwargs=dict(config=self))
INTERNALERROR>   File "/home/worldmind/.local/lib/python3.7/site-packages/pluggy/hooks.py", line 308, in call_historic
INTERNALERROR>     res = self._hookexec(self, self.get_hookimpls(), kwargs)
INTERNALERROR>   File "/home/worldmind/.local/lib/python3.7/site-packages/pluggy/manager.py", line 92, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File "/home/worldmind/.local/lib/python3.7/site-packages/pluggy/manager.py", line 86, in <lambda>
INTERNALERROR>     firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
INTERNALERROR>   File "/home/worldmind/.local/lib/python3.7/site-packages/pluggy/callers.py", line 208, in _multicall
INTERNALERROR>     return outcome.get_result()
INTERNALERROR>   File "/home/worldmind/.local/lib/python3.7/site-packages/pluggy/callers.py", line 80, in get_result
INTERNALERROR>     raise ex[1].with_traceback(ex[2])
INTERNALERROR>   File "/home/worldmind/.local/lib/python3.7/site-packages/pluggy/callers.py", line 187, in _multicall
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File "/home/worldmind/.local/lib/python3.7/site-packages/xonsh/pytest_plugin.py", line 13, in pytest_configure
INTERNALERROR>     setup()
INTERNALERROR>   File "/home/worldmind/.local/lib/python3.7/site-packages/xonsh/__amalgam__.py", line 24133, in setup
INTERNALERROR>     execer = Execer(xonsh_ctx=ctx)
INTERNALERROR>   File "/home/worldmind/.local/lib/python3.7/site-packages/xonsh/__amalgam__.py", line 23029, in __init__
INTERNALERROR>     load_builtins(execer=self, ctx=xonsh_ctx)
INTERNALERROR>   File "/home/worldmind/.local/lib/python3.7/site-packages/xonsh/__amalgam__.py", line 22760, in load_builtins
INTERNALERROR>     builtins.__xonsh__.load(execer=execer, ctx=ctx)
INTERNALERROR>   File "/home/worldmind/.local/lib/python3.7/site-packages/xonsh/__amalgam__.py", line 22832, in load
INTERNALERROR>     self.env = Env(default_env())
INTERNALERROR>   File "/home/worldmind/.local/lib/python3.7/site-packages/xonsh/__amalgam__.py", line 19255, in __init__
INTERNALERROR>     self[key] = val
INTERNALERROR>   File "/home/worldmind/.local/lib/python3.7/site-packages/xonsh/__amalgam__.py", line 19424, in __setitem__
INTERNALERROR>     self.replace_env()
INTERNALERROR>   File "/home/worldmind/.local/lib/python3.7/site-packages/xonsh/__amalgam__.py", line 19292, in replace_env
INTERNALERROR>     os_environ.update(self.detype())
INTERNALERROR>   File "/home/worldmind/.local/lib/python3.7/site-packages/xonsh/__amalgam__.py", line 19277, in detype
INTERNALERROR>     deval = ensurer.detype(val)
INTERNALERROR>   File "/home/worldmind/.local/lib/python3.7/site-packages/xonsh/__amalgam__.py", line 5498, in detype
INTERNALERROR>     return x.detype()
INTERNALERROR>   File "/home/worldmind/.local/lib/python3.7/site-packages/xonsh/__amalgam__.py", line 18217, in detype
INTERNALERROR>     style = self.style
INTERNALERROR>   File "/home/worldmind/.local/lib/python3.7/site-packages/xonsh/__amalgam__.py", line 18249, in style
INTERNALERROR>     style_name = self.style_name
INTERNALERROR>   File "/home/worldmind/.local/lib/python3.7/site-packages/xonsh/__amalgam__.py", line 18239, in style_name
INTERNALERROR>     env = builtins.__xonsh__.env
INTERNALERROR> AttributeError: 'XonshSession' object has no attribute 'env'
```
xonfig                                                                                                                        
```
+------------------+----------------------+
| xonsh            | 0.9.13.dev1          |
| Git SHA          | 9f7ccc65             |
| Commit Date      | Oct 15 17:14:50 2019 |
| Python           | 3.7.3                |
| PLY              | 3.11                 |
| have readline    | True                 |
| prompt toolkit   | 2.0.10               |
| shell type       | prompt_toolkit2      |
| pygments         | 2.4.2                |
| on posix         | True                 |
| on linux         | True                 |
| distro           | ubuntu               |
| on darwin        | False                |
| on windows       | False                |
| on cygwin        | False                |
| on msys2         | False                |
| is superuser     | False                |
| default encoding | utf-8                |
| xonsh encoding   | utf-8                |
| encoding errors  | surrogateescape      |
+------------------+----------------------+
```

**Comment by wtee** [comment]
I'm getting the same traceback whenever I try to use xonsh within tmux.

```
Traceback (most recent call last):
  File "/usr/lib/python3.8/site-packages/xonsh/__amalgam__.py", line 24015, in main
    args = premain(argv)
  File "/usr/lib/python3.8/site-packages/xonsh/__amalgam__.py", line 23968, in premain
    env = start_services(shell_kwargs, args)
  File "/usr/lib/python3.8/site-packages/xonsh/__amalgam__.py", line 23899, in start_services
    execer = Execer(
  File "/usr/lib/python3.8/site-packages/xonsh/__amalgam__.py", line 23029, in __init__
    load_builtins(execer=self, ctx=xonsh_ctx)
  File "/usr/lib/python3.8/site-packages/xonsh/__amalgam__.py", line 22760, in load_builtins
    builtins.__xonsh__.load(execer=execer, ctx=ctx)
  File "/usr/lib/python3.8/site-packages/xonsh/__amalgam__.py", line 22832, in load
    self.env = Env(default_env())
  File "/usr/lib/python3.8/site-packages/xonsh/__amalgam__.py", line 19255, in __init__
    self[key] = val
  File "/usr/lib/python3.8/site-packages/xonsh/__amalgam__.py", line 19424, in __setitem__
    self.replace_env()
  File "/usr/lib/python3.8/site-packages/xonsh/__amalgam__.py", line 19292, in replace_env
    os_environ.update(self.detype())
  File "/usr/lib/python3.8/site-packages/xonsh/__amalgam__.py", line 19277, in detype
    deval = ensurer.detype(val)
  File "/usr/lib/python3.8/site-packages/xonsh/__amalgam__.py", line 5498, in detype
    return x.detype()
  File "/usr/lib/python3.8/site-packages/xonsh/__amalgam__.py", line 18217, in detype
    style = self.style
  File "/usr/lib/python3.8/site-packages/xonsh/__amalgam__.py", line 18249, in style
    style_name = self.style_name
  File "/usr/lib/python3.8/site-packages/xonsh/__amalgam__.py", line 18239, in style_name
    env = builtins.__xonsh__.env
AttributeError: 'XonshSession' object has no attribute 'env'
```
xonfig:

```
+------------------+-----------------+
| xonsh            | 0.9.13          |
| Python           | 3.8.0           |
| PLY              | 3.11            |
| have readline    | True            |
| prompt toolkit   | 2.0.10          |
| shell type       | prompt_toolkit2 |
| pygments         | 2.4.2           |
| on posix         | True            |
| on linux         | True            |
| distro           | arch            |
| on darwin        | False           |
| on windows       | False           |
| on cygwin        | False           |
| on msys2         | False           |
| is superuser     | False           |
| default encoding | utf-8           |
| xonsh encoding   | utf-8           |
| encoding errors  | surrogateescape |
+------------------+-----------------+
```

**Comment by dyuri** [comment]
Same issue here whenever I try to execute `xonsh` from a running `xonsh` instance.
```
Traceback (most recent call last):
  File "/home/dyuri/.local/lib/python3.8/site-packages/xonsh/main.py", line 399, in main
    args = premain(argv)
  File "/home/dyuri/.local/lib/python3.8/site-packages/xonsh/main.py", line 352, in premain
    env = start_services(shell_kwargs, args)
  File "/home/dyuri/.local/lib/python3.8/site-packages/xonsh/main.py", line 283, in start_services
    execer = Execer(
  File "/home/dyuri/.local/lib/python3.8/site-packages/xonsh/execer.py", line 63, in __init__
    load_builtins(execer=self, ctx=xonsh_ctx)
  File "/home/dyuri/.local/lib/python3.8/site-packages/xonsh/built_ins.py", line 1288, in load_builtins
    builtins.__xonsh__.load(execer=execer, ctx=ctx)
  File "/home/dyuri/.local/lib/python3.8/site-packages/xonsh/built_ins.py", line 1360, in load
    self.env = Env(default_env())
  File "/home/dyuri/.local/lib/python3.8/site-packages/xonsh/environ.py", line 1402, in __init__
    self[key] = val
  File "/home/dyuri/.local/lib/python3.8/site-packages/xonsh/environ.py", line 1571, in __setitem__
    self.replace_env()
  File "/home/dyuri/.local/lib/python3.8/site-packages/xonsh/environ.py", line 1439, in replace_env
    os_environ.update(self.detype())
  File "/home/dyuri/.local/lib/python3.8/site-packages/xonsh/environ.py", line 1424, in detype
    deval = ensurer.detype(val)
  File "/home/dyuri/.local/lib/python3.8/site-packages/xonsh/tools.py", line 1103, in detype
    return x.detype()
  File "/home/dyuri/.local/lib/python3.8/site-packages/xonsh/environ.py", line 364, in detype
    style = self.style
  File "/home/dyuri/.local/lib/python3.8/site-packages/xonsh/environ.py", line 396, in style
    style_name = self.style_name
  File "/home/dyuri/.local/lib/python3.8/site-packages/xonsh/environ.py", line 386, in style_name
    env = builtins.__xonsh__.env
AttributeError: 'XonshSession' object has no attribute 'env'
Xonsh encountered an issue during launch
Failback to /bin/sh
```
xonfig:
```
+------------------+----------------------+
| xonsh            | 0.9.13.dev1          |
| Git SHA          | 9f7ccc65             |
| Commit Date      | Oct 15 17:14:50 2019 |
| Python           | 3.8.1                |
| PLY              | 3.11                 |
| have readline    | True                 |
| prompt toolkit   | 3.0.2                |
| shell type       | prompt_toolkit2      |
| pygments         | 2.5.2                |
| on posix         | True                 |
| on linux         | True                 |
| distro           | arch                 |
| on darwin        | False                |
| on windows       | False                |
| on cygwin        | False                |
| on msys2         | False                |
| is superuser     | False                |
| default encoding | utf-8                |
| xonsh encoding   | utf-8                |
| encoding errors  | surrogateescape      |
+------------------+----------------------+
```

If I delete `$LS_COLORS` it starts up fine, but in the new shell `$LS_COLORS` are defined again, even if I don't have it in any of my startup rc files (`.xonshrc`, `.bashrc`, etc.), and even if I execute `xonsh` with `--no-rc`.

**Comment by dyuri** [comment]
This small change "fixes" (or at least "hides") the issue for me: https://github.com/dyuri/xonsh/commit/fb00efccfad8a926a49f14faad2eb56b0ae70f49

**Comment by anki-code** [comment]
Bump. I've faced with this error in [xonssh/xxh](https://github.com/xonssh/xxh). After running xonsh from appimage and tryin to run xonsh xsh I've got the error as described above. And `del $LS_COLORS` fid the issue.

**Comment by scopatz** [comment]
@dyuri or @anki-code - could you put in a PR with the fix, please?

**Comment by gforsyth** [comment]
Closed via #3464 -- thanks for the fix @dyuri and thanks all for reporting!  Please ping if the issue persists on master.
